### PR TITLE
Fix race-condition on memory pool

### PR DESCRIPTION
**Pull request by sonots**
For https://github.com/cupy/cupy/issues/317#issuecomment-317614730

1. Remove in_use flag to avoid race-condition
2. Change free list from `list` to `set` so that we can find a free block fast without in_use flag
3. Add _free_lock and _in_use_lock to avoid race-condition with free_all_blocks, n_free_blocks(), used_bytes(), free_bytes(), and total_bytes().
    * Operations on `set` or `dict` are atomic in python, but locks are required for these functions which iterate set or dict.

### CODE REVIEW

**Comment by sonots** [comment]
I could not reproduce the error on my environment, so I could not write a test, but we've confirmed that the error was resolved on his environment.

**Comment by sonots** [comment]
jenkins, test this please

**Comment by sonots** [comment]
I benchmarked, and observed performance degradation.

```python
import time
import cupy
pool = cupy.cuda.MemoryPool()

m = pool.malloc(2048)
del m

start_time = time.time()
for i in range(10000):
    m1 = pool.malloc(512)
    m2 = pool.malloc(512)
    m3 = pool.malloc(512)
    m4 = pool.malloc(512)
    del m2
    del m3
    del m1
    del m4
elapsed_time = time.time() - start_time
print(elapsed_time)
```

Before: 0.10920357704162598
After: 0.17422246932983398

I will try another version with no lock, but `used_bytes()`, etc is not thread-safe.

**Comment by sonots** [comment]
No lock version: 0.12729454040527344

https://github.com/cupy/cupy/compare/master...sonots:branch_fix_race_condition_memory_pool_no_lock?expand=1

**Comment by sonots** [comment]
I tried fastrlock https://github.com/scoder/fastrlock

https://github.com/cupy/cupy/commit/1a6fda6cfc4f3ae38cb246c10a64ff3bb2f67020

But, it did not improve performance at all ...

FastRLock version: 0.20251154899597168

**Comment by scoder** [comment]
As a general comment, it's often better (and always faster) to rely on the GIL, rather than using your own locking. But I admit that I haven't looked deeply enough into this to be sure that this can be done in the case at hand.

**Comment by sonots** [comment]
I tried to call fastrlock C API  > https://github.com/sonots/cupy/commit/5449a292e83f0728bad176c91e083d79170df16a

This version drastically improved performance!

FastRLock version: 0.11914539337158203

**Comment by sonots** [comment]
rebased to master

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -492,42 +493,80 @@ cdef class SingleDeviceMemoryPool:
...
-        if chunk.next and not chunk.next.in_use:
+        if chunk.next:
+            chunk_next = None
             index = self._bin_index_from_size(chunk.next.size)
-            self._free[index].remove(chunk.next)
-            chunk = self._merge(chunk, chunk.next)
-
-        if chunk.prev and not chunk.prev.in_use:
+            try:
+                rlock.lock_fastrlock(self._free_lock, -1, True)
```
Comment (okuta) [inline cupy/cuda/memory.pyx]: How about unify `rlock.lock_fastrlock(self._free_lock, -1, True)`?
Up to three lock and free about `self._free_lock` are executed in this method.

File: cupy/cuda/memory.pyx [hunk 2]
```diff
@@ -462,14 +458,19 @@ cdef class SingleDeviceMemoryPool:
...
-            free_list = self._free[i]
-            if free_list:
-                chunk = free_list.pop()
-                chunk, _remaining = self._split(chunk, size)
-                break
-
-        # cudaMalloc if not found
-        if chunk is None:
+            try:
+                rlock.lock_fastrlock(self._free_lock, -1, True)
```
Comment (okuta) [inline cupy/cuda/memory.pyx]: How about moving to L460 (out of `for i in range(index, length)`)?

File: cupy/cuda/memory.pyx [hunk 3]
```diff
@@ -492,42 +493,80 @@ cdef class SingleDeviceMemoryPool:
...
-                if not chunk.prev and not chunk.next:
-                    free_list.remove(chunk)
+        try:
+            rlock.lock_fastrlock(self._free_lock, -1, True)
+            for i in range(len(self._free)):
+                keep_list = set()
+                for chunk in self._free[i]:
+                    if chunk.prev or chunk.next:
+                        keep_list.add(chunk)
+                self._free[i].clear()
```
Comment (okuta) [inline cupy/cuda/memory.pyx]: Please remove this line.

File: cupy/cuda/memory.pyx [hunk 4]
```diff
@@ -492,42 +493,80 @@ cdef class SingleDeviceMemoryPool:
...
+            chunk_next = None
             index = self._bin_index_from_size(chunk.next.size)
-            self._free[index].remove(chunk.next)
-            chunk = self._merge(chunk, chunk.next)
-
-        if chunk.prev and not chunk.prev.in_use:
+            try:
+                rlock.lock_fastrlock(self._free_lock, -1, True)
+                if chunk.next in self._free[index]:
+                    self._free[index].remove(chunk.next)
```
Comment (okuta) [inline cupy/cuda/memory.pyx]: How about adding type hint?
`(<set>self._free[index]).remove(chunk.next)`

File: cupy/cuda/memory.pyx (hunk 1 above)
Comment (sonots) [inline cupy/cuda/memory.pyx]: Generally speaking, lock duration should be minimized as much as possible to minimize blocking duration.
I even moved `chunk = self._merge(chunk, chunk_next)` to outside of the lock to minimize blocking duration.
I understand that you cared of overhead of calling `lock` and `unlock`.
However, I am pretty sure that minimized locking overcomes larger locking in terms of performance with multi-threaded application.

File: cupy/cuda/memory.pyx (hunk 2 above)
Comment (sonots) [inline cupy/cuda/memory.pyx]: ditto ( https://github.com/cupy/cupy/pull/382#discussion_r133328956 )

File: cupy_setup_build.py [hunk 5]
```diff
@@ -529,4 +527,7 @@ def _wrap_new_compiler(*args, **kwargs):
             # Intentionally causes DistutilsPlatformError in
             # ccompiler.new_compiler() function to hook.
             self.compiler = 'nvidia'
+        use_cython = check_cython_version()
```
Comment (sonots) [inline cupy_setup_build.py]: changed

File: cupy/cuda/memory.pyx (hunk 4 above)
Comment (sonots) [inline cupy/cuda/memory.pyx]: changed

File: cupy/cuda/memory.pyx (hunk 3 above)
Comment (sonots) [inline cupy/cuda/memory.pyx]: changed

File: cupy/cuda/memory.pyx (hunk 4 above)
Comment (sonots) [inline cupy/cuda/memory.pyx]: reverted because I did not get any changes in cpp.

File: cupy/cuda/memory.pyx (hunk 1 above)
Comment (okuta) [inline cupy/cuda/memory.pyx]: OK, I understand.

File: cupy/cuda/memory.pyx (hunk 2 above)
Comment (okuta) [inline cupy/cuda/memory.pyx]: OK, I understand.
How about using double checked locking for speed up?

Comment (sonots) [inline cupy/cuda/memory.pyx]: This lock does exclusion control with `free_all_blocks`, too.
I am not sure how we can do double-checked locking here.
Can you elaborate more?

File: cupy/cuda/memory.pyx (hunk 4 above)
Comment (okuta) [inline cupy/cuda/memory.pyx]: Oh...

File: cupy/cuda/memory.pyx (hunk 2 above)
Comment (okuta) [inline cupy/cuda/memory.pyx]: Sorry, This comment is mistake.

Comment (okuta) [inline cupy/cuda/memory.pyx]:
```python
if self._free[i]:
  try:
    rlock.lock_fastrlock(self._free_lock, -1, True)
    free_list = self._free[i]
    if free_list:
     ...
```

Comment (sonots) [inline cupy/cuda/memory.pyx]: okay, should work...
changed.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**